File: VirtualWaitingRoom/php/clinicalViewer/appointmentQuery.php
<?php
declare(strict_types=1);
#---------------------------------------------------------------------------------------------------------------
# This script finds all appointments matching the specified criteria and returns patient information from the ORMS database.
#---------------------------------------------------------------------------------------------------------------

require("../loadConfigs.php");

#get input parameters

$sDateInit = $_GET["sDate"] ?? NULL;
$eDateInit = $_GET["eDate"] ?? NULL;
$sTime = $_GET["sTime"] ?? NULL;
$eTime = $_GET["eTime"] ?? NULL;
$clinic = $_GET["clinic"] ?? NULL;
$statusConditions = [
    "completed" => isset($_GET["comp"]) ? "'Completed'" : NULL,
    "open" => isset($_GET["openn"]) ? "'Open'" : NULL,
    "cancelled" => isset($_GET["canc"]) ? "'Cancelled'" : NULL,
];

$checkForArrived = isset($_GET["arrived"]) ? TRUE : NULL;
$checkForNotArrived = isset($_GET["notArrived"]) ? TRUE : NULL;
$OPAL = isset($_GET["opal"]) ? TRUE : NULL;
$SMS = isset($_GET["SMS"]) ? TRUE : NULL;
$appType = $_GET["type"] ?? NULL;
$specificApp = $_GET["specificType"] ?? "NULL";
$appcType = $_GET["ctype"] ?? NULL;
$cspecificApp = $_GET["cspecificType"] ?? "NULL";
$appdType = $_GET["dtype"] ?? NULL;
$dspecificApp = $_GET["dspecificType"] ?? "NULL";
$checkRamqExpiration = isset($_GET["checkRamqExpiration"]) ? TRUE : NULL;
$qType = $_GET["qtype"] ?? NULL;
$qspecificApp = $_GET["qspecificType"] ?? "NULL";
$qDateInit = $_GET["qselectedDate"] ?? NULL;
$qTime = $_GET["qselectedTime"] ?? NULL;
$offbutton = $_GET["offbutton"]??NULL;

$sDate = "$sDateInit $sTime";
$eDate = "$eDateInit $eTime";
$qDate = "$qDateInit $qTime";


#query the database
$dbh = new PDO(WRM_CONNECT,MYSQL_USERNAME,MYSQL_PASSWORD,$WRM_OPTIONS);


$dbOpal = new PDO(OPAL_CONNECT,OPAL_USERNAME,OPAL_PASSWORD,$OPAL_OPTIONS);

$sqlOpal = "
    SELECT
            DT.Name_EN,
            Q.QuestionnaireCompletionDate,
            Q.CompletedWithinLastWeek,
            Q.RecentAnswered
        FROM
            (SELECT P.PatientId, D.DiagnosisCode
             from Patient P 
             INNER JOIN Diagnosis D ON D.PatientSerNum = P.PatientSerNum
             AND P.PatientId = :uid
             ORDER BY D.LastUpdated DESC
             LIMIT 1) DP
             INNER JOIN DiagnosisCode DC ON DC.DiagnosisCode = DP.DiagnosisCode
             INNER JOIN  DiagnosisTranslation DT ON DC.DiagnosisTranslationSerNum = DT.DiagnosisTranslationSerNum,
            (SELECT
                Questionnaire.CompletionDate AS QuestionnaireCompletionDate,
                CASE
                    WHEN Questionnaire.CompletionDate BETWEEN DATE_SUB(CURDATE(), INTERVAL 7 DAY) AND NOW() THEN 1
                        ELSE 0
                    END AS CompletedWithinLastWeek,
                CASE
                    WHEN Questionnaire.CompletionDate BETWEEN :qDate AND NOW() THEN 1
                        ELSE 0
                    END AS RecentAnswered
            FROM
                Patient
                INNER JOIN Questionnaire ON Questionnaire.PatientSerNum = Patient.PatientSerNum
                    AND Questionnaire.CompletedFlag = 1
                    AND Patient.PatientId = :uid2
                    LIMIT 1) Q
    LIMIT 1";

$queryOpal = $dbOpal->prepare($sqlOpal);

$sqlOpal2 = "
    SELECT DISTINCT QC.QuestionnaireName_EN AS QuestionnaireName
            FROM
                Patient
                INNER JOIN Questionnaire ON Questionnaire.PatientSerNum = Patient.PatientSerNum
                AND Questionnaire.CompletedFlag = 1
                Inner JOIN QuestionnaireControl QC ON QC.QuestionnaireControlSerNum = Questionnaire.QuestionnaireControlSerNum 
                AND Patient.PatientId = :uid";

$queryOpal2 = $dbOpal->prepare($sqlOpal2);


$specialityFilter = "ClinicResources.Speciality = '$clinic'";

$activeStatusConditions = array_filter($statusConditions);
//if($specificApp === NULL) print("t");
//else print("x");
$opalFilter = "";
if($OPAL==NULL) $opalFilter .= "AND Patient.SMSAlertNum IS NOT NULL";

if($SMS==NULL) $opalFilter .= "AND Patient.OpalPatient = 1";


$statusFilter = " AND MV.Status IN (" . implode(",", $activeStatusConditions) . ")";
$appFilter = ($appType === "all") ? "" : " AND MV.ResourceDescription IN (" . implode(",", explode(",",$specificApp)) . ")";
$cappFilter = ($appcType === "all") ? "" : " AND MV.AppointmentCode IN (" . implode(",", explode(",",$cspecificApp)) . ")";

//$appFilter = ($appType === "all") ? "" : " AND MV.ResourceDescription IN :resDesc ";
//$cappFilter = ($appcType === "all") ? "" : " AND MV.AppointmentCode IN :appCode ";

/*print_r(explode(",",$dspecificApp));
if(in_array("Sarcoma", explode(",",$dspecificApp))) print "good";

else print "bad";*/


$sql = "
    SELECT
        MV.AppointmentSerNum,
        Patient.FirstName,
        Patient.LastName,
        Patient.PatientId,
        Patient.SSN,
        MV.ResourceDescription,
        MV.Resource,
        MV.AppointmentCode,
        MV.Status,
        MV.ScheduledDate,
        MV.ScheduledTime,
        MV.CreationDate,
        MV.ReferringPhysician,
        (select PL.ArrivalDateTime from PatientLocation PL where PL.AppointmentSerNum = MV.AppointmentSerNum AND PL.PatientLocationRevCount = 1 limit 1) as CurrentCheckInTime,
        (select PLM.ArrivalDateTime from PatientLocationMH PLM where PLM.AppointmentSerNum = MV.AppointmentSerNum AND PLM.PatientLocationRevCount = 1 limit 1) as PreviousCheckInTime,
        MV.MedivisitStatus,
        (SELECT DATE_FORMAT(MAX(TEMP_PatientQuestionnaireReview.ReviewTimestamp),'%Y-%m-%d %H:%i') FROM TEMP_PatientQuestionnaireReview WHERE TEMP_PatientQuestionnaireReview.PatientSer = Patient.PatientSerNum) AS LastQuestionnaireReview,
        Patient.OpalPatient,
        Patient.SMSAlertNum
    FROM
        Patient
        INNER JOIN MediVisitAppointmentList MV ON MV.PatientSerNum = Patient.PatientSerNum
        AND MV.Status != 'Deleted'
        AND MV.ResourceDescription IN (
            SELECT DISTINCT
                ClinicResources.ResourceName
            FROM ClinicResources
            WHERE $specialityFilter
        )
        AND MV.ScheduledDateTime BETWEEN :sDate AND :eDate
        $statusFilter
        $appFilter
        $cappFilter
        $opalFilter

    ORDER BY
        MV.ScheduledDate,
        MV.ScheduledTime";


$query = $dbh->prepare($sql);

$qappFilter = ($qType === "all") ? "" : " AND QC.QuestionnaireName_EN IN ('" . implode("','", explode(",",$qspecificApp)) . "')";
$dappFilter = ($appdType === "all") ? "" : " AND DT.Name_EN IN ('" . implode("','", explode(",",$dspecificApp)) . "')";

$sqlOpal3 = "SELECT P.PatientId,
            (select DT.Name_EN from Diagnosis D
             INNER JOIN DiagnosisCode DC ON DC.DiagnosisCode = D.DiagnosisCode
             INNER JOIN DiagnosisTranslation DT ON DC.DiagnosisTranslationSerNum = DT.DiagnosisTranslationSerNum
             $dappFilter
             where D.PatientSerNum = P.PatientSerNum
             ORDER BY D.LastUpdated DESC
             LIMIT 1) AS Name_EN,
             (SELECT Q.CompletionDate
                FROM Questionnaire Q 
				Inner JOIN QuestionnaireControl QC ON QC.QuestionnaireControlSerNum = Q.QuestionnaireControlSerNum 
				AND Q.CompletedFlag = 1
				$qappFilter
				where Q.PatientSerNum = P.PatientSerNum
				ORDER BY Q.CompletionDate DESC
				LIMIT 1) AS QuestionnaireCompletionDate,
			 (SELECT QC.QuestionnaireName_EN 
                FROM Questionnaire Q 
				Inner JOIN QuestionnaireControl QC ON QC.QuestionnaireControlSerNum = Q.QuestionnaireControlSerNum 
				AND Q.CompletedFlag = 1
				$qappFilter
				where Q.PatientSerNum = P.PatientSerNum
				ORDER BY Q.CompletionDate DESC
				LIMIT 1) AS QuestionnaireName,
			(SELECT CASE
                    WHEN Q.CompletionDate BETWEEN DATE_SUB(CURDATE(), INTERVAL 7 DAY) AND NOW() THEN 1
                        ELSE 0
                    END AS CompletedWithinLastWeek
                FROM Questionnaire Q 
				Inner JOIN QuestionnaireControl QC ON QC.QuestionnaireControlSerNum = Q.QuestionnaireControlSerNum 
				AND Q.CompletedFlag = 1
				$qappFilter
				where Q.PatientSerNum = P.PatientSerNum
				ORDER BY Q.CompletionDate DESC
				LIMIT 1) AS CompletedWithinLastWeek,
			(SELECT CASE
                    WHEN Q.CompletionDate BETWEEN :qDate AND NOW() THEN 1
                        ELSE 0
                    END AS RecentAnswered
                FROM Questionnaire Q 
				Inner JOIN QuestionnaireControl QC ON QC.QuestionnaireControlSerNum = Q.QuestionnaireControlSerNum 
				AND Q.CompletedFlag = 1
				$qappFilter
				where Q.PatientSerNum = P.PatientSerNum
				ORDER BY Q.CompletionDate DESC
				LIMIT 1) AS RecentAnswered
            FROM Patient P";

$queryOpal3 = $dbOpal->prepare($sqlOpal3);

$sql2 = "SELECT  Patient.FirstName,
        Patient.LastName,
        Patient.PatientId,
        Patient.SSN,
        (SELECT DATE_FORMAT(MAX(TEMP_PatientQuestionnaireReview.ReviewTimestamp),'%Y-%m-%d %H:%i') FROM TEMP_PatientQuestionnaireReview WHERE TEMP_PatientQuestionnaireReview.PatientSer = Patient.PatientSerNum) AS LastQuestionnaireReview,
        Patient.OpalPatient,
        Patient.SMSAlertNum
        FROM Patient
        WHERE Patient.PatientId = :uid";

$query2 = $dbh->prepare($sql2);

//if ($appFilter !== "") $query->bindValue(":resDesc", $specificApp);
//if ($cappFilter !== "") $query->bindValue(":appCode", $cspecificApp);
$query->bindValue(":sDate", $sDate);
$query->bindValue(":eDate", $eDate);

$query->execute();

$listOfAppointments = [];
$pIDlist = [];

while ($row = $query->fetch(PDO::FETCH_ASSOC)) {
    #filter apppointments on whether the patient checked in for it
    $checkInTime = $row["CurrentCheckInTime"] ?? $row["PreviousCheckInTime"] ?? NULL;

    if ($checkForArrived === TRUE && $checkForNotArrived === TRUE) $filterAppointment = FALSE;
    elseif ($checkForArrived === TRUE && $checkInTime === NULL) $filterAppointment = TRUE;
    elseif ($checkForNotArrived === TRUE && $checkInTime !== NULL) $filterAppointment = TRUE;
    else $filterAppointment = FALSE;

    if ($filterAppointment) continue;

    #if ramq expiration checking is enabled, check if the ramq is expired
    /*
        $ramqExpired = FALSE;
        if ($checkRamqExpiration) {
            $ramqInfo = AdtWebservice::getRamqInformation($row["SSN"]);
            $row["ssnExp"] = $ramqInfo["Expiration"];
            if ($ramqInfo["Status"] === "Expired" || $ramqInfo["Status"] === "Error") $ramqExpired = TRUE;
        }
    */

    #perform some processing
    if (isset($row["ssnExp"]) && $row["ssnExp"] !== NULL) $row["ssnExp"] = (new DateTime($row["ssnExp"]))->format("ym");
    $row["ScheduledTime"] = substr($row["ScheduledTime"], 0, -3);

    $queryOpal->execute(array(":uid" => $row["PatientId"],":qDate" => $qDate,":uid2" => $row["PatientId"]));
    $resultOpal = $queryOpal->fetch(PDO::FETCH_ASSOC);

    $row["Diagnosis"] = !empty($resultOpal["Name_EN"]) ? $resultOpal["Name_EN"] : "";
    $row["QuestionnaireName"] = !empty($resultOpal["QuestionnaireName"]) ? $resultOpal["QuestionnaireName"] : "";
    if($row["SMSAlertNum"]) $row["SMSAlertNum"] = substr($row["SMSAlertNum"],0,3) ."-". substr($row["SMSAlertNum"],3,3) ."-". substr($row["SMSAlertNum"],6,4);

    $lastCompleted = $resultOpal["QuestionnaireCompletionDate"] ?? NULL;
    $completedWithinWeek = $resultOpal["CompletedWithinLastWeek"] ?? NULL;
    $row["QStatus"] = ($completedWithinWeek === "1") ? "green-circle" : "";

    if(
        ($lastCompleted !== NULL && $row["LastQuestionnaireReview"] === NULL)
        ||
        (
            ($lastCompleted !== NULL && $row["LastQuestionnaireReview"] !== NULL)
            && (new DateTime($lastCompleted))->getTimestamp() > (new DateTime($row["LastQuestionnaireReview"]))->getTimestamp()
        )
    ) $row["QStatus"] = "red-circle";

    $answeredQuestionnaire = False;
    $queryOpal2->execute(array(":uid" => $row["PatientId"]));
    while ($questionnaireName = $queryOpal2->fetch(PDO::FETCH_ASSOC)){
        if(in_array( $questionnaireName["QuestionnaireName"],explode(",",$qspecificApp))){
            $answeredQuestionnaire = true;
            break;
        }
    }

    if(!in_array($row["PatientId"], $pIDlist)){
        array_push($pIDlist,$row["PatientId"]);
    }

    if(($appdType ==="all" || in_array( $row["Diagnosis"],explode(",",$dspecificApp))) && ($resultOpal["RecentAnswered"]==1||$offbutton == "OFF")&&($qType ==="all" ||$answeredQuestionnaire)){
        $listOfAppointments[] = [
            "fname" => $row["FirstName"],
            "lname" => $row["LastName"],
            "pID" => $row["PatientId"],
            "ssn" => [
                "num" => $row["SSN"],
                "expDate" => $row["ssnExp"] ?? NULL,
                "expired" => $ramqExpired,
            ],
            "appName" => $row["ResourceDescription"],
            "appClinic" => $row["Resource"],
            "appType" => $row["AppointmentCode"],
            "appStatus" => $row["Status"],
            "appDay" => $row["ScheduledDate"],
            "appTime" => $row["ScheduledTime"],
            "checkin" => $checkInTime,
            "createdToday" => new DateTime($row["CreationDate"]) == new DateTime("midnight"),
            "referringPhysician" => $row["ReferringPhysician"],
            "mediStatus" => $row["MedivisitStatus"],
            "diagnosis" => $row["Diagnosis"],
            "QStatus" => $row["QStatus"],
            "opalpatient" =>$row["OpalPatient"],
            "SMSAlertNum" => $row["SMSAlertNum"],
        ];
    }
}

$queryOpal3->bindValue(":qDate", $qDate);

$queryOpal3->execute();

while ($row2 = $queryOpal3->fetch(PDO::FETCH_ASSOC)) {
    $query2->execute(array(":uid" => $row2["PatientId"]));
    $resultORMS = $query2->fetch(PDO::FETCH_ASSOC);

    if($resultORMS["SMSAlertNum"]) $resultORMS["SMSAlertNum"] = substr($resultORMS["SMSAlertNum"],0,3) .
        "-". substr($resultORMS["SMSAlertNum"],3,3) ."-". substr($resultORMS["SMSAlertNum"],6,4);

    $lastCompleted = $row2["QuestionnaireCompletionDate"] ?? NULL;
    $completedWithinWeek = $row2["CompletedWithinLastWeek"] ?? NULL;
    $row2["QStatus"] = ($completedWithinWeek === "1") ? "green-circle" : "";

    if(
        ($lastCompleted !== NULL && $resultORMS["LastQuestionnaireReview"] === NULL)
        ||
        (
            ($lastCompleted !== NULL && $resultORMS["LastQuestionnaireReview"] !== NULL)
            && (new DateTime($lastCompleted))->getTimestamp() > (new DateTime($resultORMS["LastQuestionnaireReview"]))->getTimestamp()
        )
    ) $row2["QStatus"] = "red-circle";


    if(($appdType ==="all" || $row2["Name_EN"]) &&($row2["RecentAnswered"]==1||$offbutton == "OFF")&&($qType ==="all" ||$row2["QuestionnaireName"]) && (!in_array($row2["PatientId"], $pIDlist))&& ($resultORMS["FirstName"])){
        $listOfAppointments[] = [
            "fname" => $resultORMS["FirstName"],
            "lname" => $resultORMS["LastName"],
            "pID" => $row2["PatientId"],
            "ssn" => [
                "num" => $row2["SSN"],
                "expDate" => $row2["ssnExp"] ?? NULL,
                "expired" => $ramqExpired,
            ],
            "appName" => $row2["ResourceDescription"],
            "appClinic" => $row2["Resource"],
            "appType" => $row2["AppointmentCode"],
            "appStatus" => $row2["Status"],
            "appDay" => $row2["ScheduledDate"],
            "appTime" => $row2["ScheduledTime"],
            "checkin" => NULL,
            "createdToday" => NULL,
            "referringPhysician" => $row2["ReferringPhysician"],
            "mediStatus" => $row2["MedivisitStatus"],
            "diagnosis" => $row2["Name_EN"],
            "QStatus" => $row2["QStatus"],
            "opalpatient" =>$resultORMS["OpalPatient"],
            "SMSAlertNum" => $resultORMS["SMSAlertNum"],
        ];
    }
}


$listOfAppointments = utf8_encode_recursive($listOfAppointments);
echo json_encode($listOfAppointments);


exit;

?>

File: reports/Scripts/AppointmentQuery.pl
#!/usr/bin/perl
#---------------------------------------------------------------------------------------------------------------
# This script finds all appointments matching the specified criteria and returns patient information from the phpmyadmin WaitRoomManagment database.
#---------------------------------------------------------------------------------------------------------------

#----------------------------------------
#import modules
#----------------------------------------
use strict;
use v5.26;

use lib '../../perl/system/modules';
use HospitalADT;
use LoadConfigs;

use CGI qw(:standard);
use CGI::Carp qw(fatalsToBrowser);
use JSON;
use Time::Piece;
use Encode;

#-----------------------------------------
#start html feedback
#-----------------------------------------
my $cgi = CGI->new;
print $cgi->header('application/json');

#------------------------------------------
#parse input parameters
#------------------------------------------
my $sDateInit = param("sDate");
my $eDateInit = param("eDate");
my $sTime = param("sTime");
my $eTime = param("eTime");
my $comp = param("comp");
my $open = param("openn");
my $prog = param('prog');
my $canc = param("canc");
my $arrived = param("arrived");
my $notArrived = param("notArrived");
my $clinic = param("clinic");
my $appType = param("type");
my $specificType = param("specificType");
my $updateRamq = param("updateRamq");

my $sDate = $sDateInit ." $sTime";
my $eDate = $eDateInit ." $eTime";

my $appFilter = "";
$appFilter .= "AND ( ";
$appFilter .= "MV.Status = 'Completed' OR " if($comp);
$appFilter .= "MV.Status = 'Open' OR " if($open);
$appFilter .= "MV.Status = 'Cancelled' OR " if($canc);
$appFilter .= "MV.Status = 'In Progress' OR " if($prog);
chop($appFilter) for (1..4);
$appFilter .=  ") ";

($specificType = $specificType) =~ s/'/\'/g;

my $specialityFilter = "AND CR.Speciality = '$clinic' ";

#escape any ' characters we find
($specificType = $specificType) =~ s/'/\\'/g;

my $typeFilter = "" if($appType eq 'all');
$typeFilter = "AND MV.ResourceDescription = '$specificType' " if($appType eq 'specific');

# print $appFilter . "\n\n";
# print $typeFilter . "\n\n";

#setup global variables
#-----------------------------------------------------
#the appointment serial will be used as the key
my %fname;
my %lname;
my %pID;
my %ssn;
my %ssnExp; #ssn expiry date
my %app; #appointment description
my %appType; #appointment type
my %appCode; #appointment code
my %appStatus; #appointment status
my %appTime; #scheduled time of appointment
my %appDay; #day of the appointment
my %creationDate; #creation date of the appointment
my %referringPhysician; #referring physician of the appointment
my %checkin; #time of patient checkin
my %mediStatus; #status of appointment in medivisit

my $format = '%Y-%m-%d %H:%M:%S';

my $jstring = {};

#-----------------------------------------------------
#connect to database and run queries
#-----------------------------------------------------
my $dbh = LoadConfigs::GetDatabaseConnection("ORMS") or die("Couldn't connect to database: ");

#get the list of possible appointments and their resources
my $sql0 = "
	SELECT DISTINCT
		MV.ResourceDescription,
		MV.Resource
	FROM
		MediVisitAppointmentList MV,
		ClinicResources CR
	WHERE
		MV.ClinicResourcesSerNum = CR.ClinicResourcesSerNum
		$specialityFilter
	ORDER BY MV.ResourceDescription,MV.Resource";

 #print $sql0 . "\n\n";

my $query0 = $dbh->prepare_cached($sql0) or die("Query could not be prepared: ".$dbh->errstr);
$query0->execute() or die("Query execution failed: ".$query0->errstr);

my $clinics = [];
my %rDesc;

while(my @data0 = $query0->fetchrow_array()) {push @{$rDesc{$data0[0]}}, $data0[1];}

foreach my $desc (sort keys %rDesc)
{
	if(scalar @{$rDesc{$desc}} > 1)
	{
		unshift @{$rDesc{$desc}}, 'All';
	}

	#remove new lines
	$desc =~ s/\n|\r//g;

    push $clinics->@*, {name=> $desc, resources=> $rDesc{$desc}->@*};
}

$jstring->{"clinics"} = $clinics;
$jstring->{"tableData"} = [];

my $sql1 = "
	SELECT
		MV.AppointmentSerNum,
		Patient.FirstName,
		Patient.LastName,
		Patient.PatientId,
		Patient.SSN,
		Patient.SSNExpDate,
		MV.ResourceDescription,
		MV.Resource,
		MV.AppointmentCode,
		MV.Status,
		MV.ScheduledDate AS ScheduledDate,
		MV.ScheduledTime AS ScheduledTime,
		MV.CreationDate,
		MV.ReferringPhysician,
		(select PL.ArrivalDateTime from PatientLocation PL where PL.AppointmentSerNum = MV.AppointmentSerNum AND PL.PatientLocationRevCount = 1 limit 1) as ArrivalDateTime,
		(select PLM.ArrivalDateTime from PatientLocationMH PLM where PLM.AppointmentSerNum = MV.AppointmentSerNum AND PLM.PatientLocationRevCount = 1 limit 1) as ArrivalDateTime,
		MV.MedivisitStatus
	FROM
		Patient,
		MediVisitAppointmentList MV
	WHERE
		Patient.PatientSerNum = MV.PatientSerNum
		AND Patient.PatientId != '9999996'
		AND Patient.PatientId != '9999998'
		AND MV.ResourceDescription in (Select distinct CR.ResourceName from ClinicResources CR Where trim(CR.ResourceName) not in ('', 'null') $specialityFilter)
		$appFilter
		AND MV.Status != 'Deleted'
		AND MV.ResourceDescription NOT LIKE '%blood%'
		AND MV.ScheduledDateTime BETWEEN '$sDate' AND '$eDate'
		$typeFilter
	ORDER BY ScheduledDate,ScheduledTime";

my $query1 = $dbh->prepare_cached($sql1) or die("Query could not be prepared: ".$dbh->errstr);
$query1->execute() or die("Query execution failed: ".$query1->errstr);

while(my @data1 = $query1->fetchrow_array())
{
	my $ser = $data1[0];
	$fname{$ser} = $data1[1];
	$lname{$ser} = $data1[2];
	$pID{$ser} = $data1[3];
	$ssn{$ser} = $data1[4];
	$ssnExp{$ser} = $data1[5];
	$ssnExp{$ser} = "0$ssnExp{$ser}" if(length($ssnExp{$ser}) eq 3);
	$app{$ser} = $data1[6];
	$appType{$ser} = $data1[7];
	$appCode{$ser} = $data1[8];
	$appStatus{$ser} = $data1[9];
	$appDay{$ser} = $data1[10];
	$appTime{$ser} = substr($data1[11],0,-3);
	$creationDate{$ser} = $data1[12];
	$referringPhysician{$ser} = $data1[13];
	$checkin{$ser} = $data1[14];
	$checkin{$ser} = $data1[15] if($data1[15]);
	$mediStatus{$ser} = $data1[16];

	#check if the ssn is expired
	my $expired = 1;

	#get information from the hospital ADT for the ramq if enabled
	#also update expired ramqs if any are found
	if($updateRamq)
	{
		my $currentDate  = localtime;

		#check if the ramq is still valid before proceeding
		#this is to reduce computation time
		if(length($ssnExp{$ser}) eq 4 and substr($ssnExp{$ser},-2) <= 12)
		{
			my $expTP = Time::Piece->strptime("20$ssnExp{$ser}","%Y%m");
			$expTP = $expTP + Time::Piece->ONE_MONTH;

			$expired = 0 if($expTP > $currentDate);
		}

		if($expired eq 1)
		{
			my $ramqInfo = HospitalADT->getRamqInformation($ssn{$ser});

			if($ramqInfo->{'Status'} =~ /Valid/)
			{
				$expired = 0;

				#also make sure the ramq in the WRM db is up to date if the ramq is expired in the db
				HospitalADT->updateRamqInWRM($ssn{$ser});
			}
		}
	}

	#set 'today' as the appointment's scheduled date so that all verifications are done as if it were the day of the appointment
	my $today = Time::Piece->strptime($appDay{$ser},'%Y-%m-%d') if($appDay{$ser});

	#check if the appointment was created today
	my $createdToday = 0;
	if($creationDate{$ser} and $creationDate{$ser} ne '0000-00-00' and $today)
	{
		my $creationTP = Time::Piece->strptime($creationDate{$ser},"%Y-%m-%d");

		$createdToday = 1 if($creationTP eq $today);
	}

	#encode the information in json format
	if(($arrived and !$notArrived and $checkin{$ser})
		or (!$arrived and $notArrived and !$checkin{$ser})
		or ($arrived and $notArrived))
	{
		#you can see that some variables appear to be in the wrong place (like appType: $appCode)
		#this is because the physicians name things different from the DB column names
        my $appObj = {
            fname=> $fname{$ser},
            lname=> $lname{$ser},
            pID=> $pID{$ser},
            ssn=> {num=> $ssn{$ser},
            expDate=> $ssnExp{$ser},
            expired=> $expired},
            appName=> $app{$ser},
            appClinic=> $appType{$ser},
            appType=> $appCode{$ser},
            appStatus=> $appStatus{$ser},
            appDay=> $appDay{$ser},
            appTime=> $appTime{$ser},
            checkin=> $checkin{$ser},
            createdToday=> $createdToday,
            referringPhysician=> $referringPhysician{$ser},
            mediStatus=> $mediStatus{$ser}
        };

        #$_ = Encode::decode('utf8',$_) for(values $appObj->%*);
		push $jstring->{"tableData"}->@*, $appObj;
	}
}

my $json = JSON->new->ascii->allow_nonref;
my $output = $json->encode($jstring);
say $output;

exit;
